responsemanager/queryexecutor/queryexecutor.go

```diff
@@ -0,0 +1,300 @@
+}
+
+// QueryExecutor is responsible for performing individual requests by executing their traversals
+type QueryExecutor struct {
+	ctx                context.Context
+	manager            Manager
+	blockHooks         BlockHooks
+	updateHooks        UpdateHooks
+	cancelledListeners CancelledListeners
+	responseAssembler  ResponseAssembler
+	workSignal         chan struct{}
+	connManager        network.ConnManager
+}
+
+// New creates a new QueryExecutor
+func New(ctx context.Context,
+	manager Manager,
+	blockHooks BlockHooks,
+	updateHooks UpdateHooks,
+	cancelledListeners CancelledListeners,
+	responseAssembler ResponseAssembler,
+	workSignal chan struct{},
+	connManager network.ConnManager,
+) *QueryExecutor {
+	qm := &QueryExecutor{
+		blockHooks:         blockHooks,
+		updateHooks:        updateHooks,
+		cancelledListeners: cancelledListeners,
+		responseAssembler:  responseAssembler,
+		manager:            manager,
+		ctx:                ctx,
+		workSignal:         workSignal,
+		connManager:        connManager,
+	}
+	return qm
+}
+
+// ExecuteTask takes a single task and executes its traversal it describes. For each block, it
+// checks for signals on the task's ResponseSignals, updates on the QueryExecutor's UpdateHooks,
+// and uses the ResponseAssembler to build and send a response, while also triggering any of
+// the QueryExecutor's BlockHooks. Traversal continues until complete, or a signal or hook
+// suggests we should stop or pause.
+func (qe *QueryExecutor) ExecuteTask(ctx context.Context, pid peer.ID, task *peertask.Task) bool {
+	// StartTask lets us block until this task is at the top of the execution stack
+	responseTaskChan := make(chan ResponseTask)
+	var rt ResponseTask
+	qe.manager.StartTask(task, responseTaskChan)
+	select {
+	case rt = <-responseTaskChan:
+	case <-qe.ctx.Done():
+		return true
+	}
+	if rt.Empty {
+		log.Info("Empty task on peer request stack")
+		return false
+	}
+
+	log.Debugw("beginning response execution", "id", rt.Request.ID(), "peer", pid.String(), "root_cid", rt.Request.Root().String())
+	err := qe.executeQuery(pid, rt)
+	isCancelled := err != nil && ipldutil.IsContextCancelErr(err)
+	if isCancelled {
+		qe.connManager.Unprotect(pid, rt.Request.ID().Tag())
+		qe.cancelledListeners.NotifyCancelledListeners(pid, rt.Request)
+	}
+	qe.manager.FinishTask(task, err)
+	log.Debugw("finishing response execution", "id", rt.Request.ID(), "peer", pid.String(), "root_cid", rt.Request.Root().String())
+	return false
+}
+
+func (qe *QueryExecutor) executeQuery(
+	p peer.ID, rt ResponseTask) error {
+
+	// Execute the traversal operation, continue until we have reason to stop (error, pause, complete)
+	err := qe.runTraversal(p, rt)
+
+	// Close out the response, either temporarily (pause) or permanently (cancel, fail, complete)
+	return qe.responseAssembler.Transaction(p, rt.Request.ID(), func(rb responseassembler.ResponseBuilder) error {
+		var code graphsync.ResponseStatusCode
+		if err != nil {
```

Comment:
sidebar, not for this PR -- I think we should make this a switch statement on err -- it would probably make for less complicated to understand code.